traffic_router/core/src/main/java/com/comcast/cdn/traffic_control/traffic_router/core/router/TrafficRouter.java

```diff
@@ -675,17 +678,40 @@ public CacheLocation getCoverageZoneCacheLocation(final String ip, final String
 			return cacheLocation;
 		}
 
-		// We had a hit in the CZF but the name does not match a known cache location.
-		// Check whether the CZF entry has a geolocation and use it if so.
-		return getClosestCacheLocation(cacheRegister.filterAvailableLocations(deliveryServiceId), networkNode.getGeolocation(), cacheRegister.getDeliveryService(deliveryServiceId));
+		if (networkNode.getBackupCacheGroups() != null) {
+			for (final String cacheGroup : networkNode.getBackupCacheGroups()) {
+				final CacheLocation bkCacheLocation = getCacheRegister().getCacheLocationById(cacheGroup);
+				if (bkCacheLocation != null && !getSupportingCaches(bkCacheLocation.getCaches(), deliveryService).isEmpty()) {
+					LOGGER.debug("Got backup CZ cache group " + bkCacheLocation.getId() + " for " + ip + ", ds " + deliveryServiceId);
+					if (track != null) {
+						track.setFromBackupCzGroup(true);
+					}
+					return bkCacheLocation;
+				}
+			}
+		} 
+
+		if (networkNode.isUseClosest()) {
+			// We had a hit in the CZF but the name does not match a known cache location.
+			// Check whether the CZF entry has a geolocation and use it if so.
+			final CacheLocation closestCacheLocation = getClosestCacheLocation(cacheRegister.filterAvailableLocations(deliveryServiceId), networkNode.getGeolocation(), cacheRegister.getDeliveryService(deliveryServiceId));
+			if (closestCacheLocation != null) {
+				LOGGER.debug("Got closest CZ cache group " + closestCacheLocation.getId() + " for " + ip + ", ds " + deliveryServiceId);
+				if (track != null) {
+					track.setFromBackupCzGroup(true);
+				}
+			}
+			return closestCacheLocation;
+		}
+		return null;
```

Comment:
Rather just returning `null` here, I think we have to return more information to the caller so that the caller knows the difference between "null because there wasn't a hit in the CZF" and "null because there _was_ a hit in the CZF but we're not returning a CacheLocation due to the `"backupZones"` config.

Reply:
@rawlinp This test case and results looks fine to me. With the Geo Limit any value other than CZF only , say for example None, TR will go all out to find a route and when doing so it will find a route via 'GEO' since this is not limited to CZF only.

For Null returning, may be we can add one more rdtl field to denote backup Zone failure. But then , i think it is supernumerary since we have enough tagging in the access logs to differentiate between selecting caches from CZ, CZ's backup configuration for successful cases.

Reply:
I disagree, one of the reasons for specifying backupZones is to confine traffic to certain cachegroups when the matched cachegroups are unavailable. In this case we're getting a hit in the CZ (so we know where the client is in the network), that cachegroup is unavailable, but the client will still fall back to a Geo-lookup (even though from the CZ-hit we already know where the client is), best-case scenario get matched to the same cachegroup, find the closest available cachegroup, and altogether disregard the "confine traffic to certain cachegroups" rule in the CZF.

Reply:
"confine traffic to certain cache group" is defined by CZF. So, if we remove that limitation (Geo Limit's value other than CZF only), TR can fetch a matching cache using Geo lookup. I am still not clear on why should TR reject this when the configuration says None as Geo Limit.

Reply:
Geo is generally the backup case when a client's IP cannot be found in the CZF. If the client is found in the CZF, we shouldn't be falling back to a Geo-lookup which subverts the CZF's `backupZones` config. We know the client is in our network, so we want it to follow the rules specified in the CZF.

If the client was _not_ found in the CZF, then it should most likely _not_ be in our network. In that case, the congestion might not be a problem because the client should be entering our network from somewhere else.

Reply:
If the client was found in the CZF, but the caches in that zone are down (and theres no backupList AND fallthrough == true), then we should try to use the`coordinates` of the matching zone to find the next closest cache group. If there are no coordinates, I think the current behavior is to try geolocation, even on a client that was in the CZF.

Reply:
I agree with that, but the scenario I'm concerned about is when `fallbackToNextClosest == false`. In that case I don't think the request should fallback to Geo when the client was found in the CZF but the matched cachegroup has no caches available. In the best-case geo-lookup that client would still be traversing the same network path to the next available cachegroup (which is what the backup list is supposed to prevent).

Reply:
ok i agree that when `fallbackToNextClosest ==false` that we should NOT geolocate

Reply:
So, fallbackToNextClosest flag of backupZones gets to decide whether TR has to fallback to Geo lookup or not. Thank you @rawlinp and @limited for clarifying.  Let me make the changes.